test/dns-test.js

```diff
@@ -0,0 +1,396 @@
+'use strict';
+
+const assert = require('bsert');
+const {NodeClient, WalletClient} = require('hs-client');
+const StubResolver = require('bns/lib/resolver/stub');
+const bcrypto = require('bcrypto');
+const dnssec = require('bns/lib/dnssec');
+const util = require('bns/lib/util');
+const wire = require('bns/lib/wire');
+const FullNode = require('../lib/node/fullnode');
+const Resource = require('../lib/dns/resource');
+const Network = require('../lib/protocol/network');
+const rules = require('../lib/covenants/rules');
+const common = require('./util/common');
+const {types} = wire;
+
+const json = require('./data/resources-v0.json');
+
+const network = Network.get('regtest');
+
+const node = new FullNode({
+  network: 'regtest',
+  apiKey: 'foo',
+  walletAuth: true,
+  rsNoUnbound: true,
+  memory: true,
+  workers: true,
+  plugins: [require('../lib/wallet/plugin')]
+});
+
+const nstub = new StubResolver({
+  rd: true,
+  cd: true,
+  edns: true,
+  ednsSize: 4096,
+  maxAttempts: 2,
+  maxTimeout: 3000,
+  dnssec: true,
+  servers: [`127.0.0.1:${network.nsPort}`]
+});
+
+const rstub = new StubResolver({
+  rd: true,
+  cd: true,
+  edns: true,
+  ednsSize: 4096,
+  maxAttempts: 2,
+  maxTimeout: 3000,
+  dnssec: true,
+  servers: [`127.0.0.1:${network.rsPort}`]
+});
+
+const nclient = new NodeClient({
+  port: network.rpcPort,
+  apiKey: 'foo'
+});
+
+const wclient = new WalletClient({
+  port: network.walletPort,
+  apiKey: 'foo'
+});
+
+const wallet = wclient.wallet('primary');
+
+const {
+  treeInterval,
+  biddingPeriod,
+  revealPeriod
+} = network.names;
+
+// miner controlled address
+let coinbase;
+
+describe('DNS Servers', function() {
+  this.timeout(15000);
+
+  before(async () => {
+    await node.open();
+    await nclient.open();
+    await wclient.open();
+    await nstub.open();
+    await rstub.open();
+
+    // mine a bunch of blocks
+    const info = await wallet.createAddress('default');
+    coinbase = info.address;
+
+    for (let i = 0; i < 15; i++)
+      await nclient.execute('generatetoaddress', [2, coinbase]);
+  });
+
+  after(async () => {
+    await nclient.close();
+    await wclient.close();
+    await node.close();
+    await nstub.close();
+    await rstub.close();
+  });
+
+  describe('Recursive Resolver', () => {
+    it('should return an A record', async () => {
+      const name = 'google.com.';
+
+      const res = await rstub.lookup(name, types.A);
```

Comment:
Is this the first time a test in bcoin or hsd actually makes an external internet connection? I wonder what the implications of that are, and what kind of failures could happen with CI. I think `google.com` seems like it will be a safe bet to query for a long time, but it still makes this test rely on an external factor. Do we want to consider a separate testing script like what I did here: https://github.com/bcoin-org/bcoin/pull/759 -- putting the test that requires an internet connection outside the main testing suite, but still running it on CI. Just 🍕 for 🤔 .

Reply:
Would agree that removing it is probably a good idea, I added it as the first test in this file to figure out how the APIs work together. Thanks for the feedback. It is a risk of nondeterminism, but if `google.com` doesn't resolve (assuming no CI firewall) there is a major problem happening with the internet

Reply:
Both `"should return no answer..."` tests also hit the internet. Can be seen by logging the queries in `bns/lib/resolver/dns.js` `exchange()`

Reply:
```
Querying server: 127.0.0.1 (43708) (tcp=false)
[ { name: 'com.', class: 'IN', type: 'A' } ]
Querying server: 192.112.36.4 (18481) (tcp=false)
[ { name: 'com.', class: 'IN', type: 'NS' } ]
      ✓ should return no answer for an ICANN rooted name (125ms)
Querying server: 127.0.0.1 (47554) (tcp=false)
[ { name: 'foo.', class: 'IN', type: 'A' } ]
Querying server: 202.12.27.33 (24311) (tcp=false)
[ { name: 'foo.', class: 'IN', type: 'NS' } ]
      ✓ should return no answer for HNS name pre-update (195ms)
```

Reply:
Going to remove all external requests. Thanks for the feedback

Reply:
Removed this external requests.